config/src/main/java/org/springframework/security/config/annotation/web/configuration/OAuth2ClientConfiguration.java

```diff
@@ -46,19 +50,34 @@
  * @since 5.1
  * @see OAuth2ImportSelector
  */
-@Import(OAuth2ClientConfiguration.OAuth2ClientWebMvcImportSelector.class)
+@Import({OAuth2ClientConfiguration.OAuth2ClientSecurityConfiguration.class,
+		OAuth2ClientConfiguration.OAuth2ClientWebMvcImportSelector.class})
 final class OAuth2ClientConfiguration {
 
+	@Configuration(proxyBeanMethods = false)
+	static class OAuth2ClientSecurityConfiguration {
+
+		@Bean
+		BeanDefinitionRegistryPostProcessor oauth2ClientRestOperationsPostProcessor() {
+			return new OAuth2ClientRestOperationsPostProcessor();
+		}
+
+		@Bean
+		BeanDefinitionRegistryPostProcessor defaultOAuth2AuthorizedClientManagerPostProcessor() {
+			return new DefaultOAuth2AuthorizedClientManagerPostProcessor();
```

Comment:
I'm having trouble seeing the value of registering `OAuth2AuthorizedClientManager` with a known bean name.

Doing so with `RestOperations` makes sense since there are many use cases when an application publishes a `RestOperations` but doesn't intend it to be used by Spring Security.

What is it about `OAuth2AuthorizedClientManager` that requires the same treatment as `RestOperations`? Is this a needed pattern for other OAuth 2.0 Client beans?

Reply:
The `OAuth2AuthorizedClientManager` is the central interface for client since 5.2. It is used primarily by `ServerOAuth2AuthorizedClientExchangeFilterFunction` and `OAuth2AuthorizedClientArgumentResolver`. It indirectly uses `RestOperations`. The relationship chain is as follows:

`OAuth2AuthorizedClientManager` has 1 or more `OAuth2AuthorizedClientProvider`
`OAuth2AuthorizedClientProvider` has an `OAuth2AccessTokenResponseClient`
`OAuth2AccessTokenResponseClient` has a `RestOperations`

If an application configures a custom `RestOperations` then it will automatically be configured with the "default" `OAuth2AuthorizedClientManager` and associated `OAuth2AuthorizedClientProvider`'s, and the user could leverage this default `@Bean` instead of explicitly configuring it. It really is meant to be a convenience mechanism.

Reply:
One important thing to note is that an application may configure more than 1 `OAuth2AuthorizedClientManager` in their setup. For example, they may have a background service that performs `client_credentials` grant and another `OAuth2AuthorizedClientManager` that handles `authorization_code` and `refresh_token` in the web context.

The "default" `OAuth2AuthorizedClientManager` will be wired into the `OAuth2AuthorizedClientArgumentResolver` and may be used by `ServletOAuth2AuthorizedClientExchangeFilterFunction`.